File: 20.valid-parentheses.cpp
/*
 * @lc app=leetcode id=20 lang=cpp
 *
 * [20] Valid Parentheses
 */

// @lc code=start
class Solution {
public:
    bool isValid(string s) {

        stack<char>st;

        for(char i: s)
        {
            if(i=='('||i=='{'||i=='[')
            {
                st.push(i);
            }
            else
            {
                if(st.empty())
                {
                    return false;
                }
                else if(i==')'&&st.top()!='(')
                {
                    return false;
                }
                else if(i=='}'&&st.top()!='{')
                {
                    return false;
                }
                 else if(i==']'&&st.top()!='[')
                {
                 return false;
                }

                st.pop();
            }
           

            
        }

        if(st.empty())
        {
            return true;
        }
        else
        {
            return false;
        }
        
    }
};
// @lc code=end

File: 20.valid-parentheses.java

import java.util.ArrayDeque;

/*
 * @lc app=leetcode id=20 lang=java
 *
 * [20] Valid Parentheses
 */

// @lc code=start
class Solution {
    public boolean isValid(String s) {

        ArrayDeque<Character>st=new ArrayDeque<>();
        for(int i=0;i<s.length();i++){
            char curr=s.charAt(i);
            if(curr=='('||curr=='{'||curr=='['){
                st.push(curr);
            }else{
                
                if(st.isEmpty()){
                    return false;
                }
                char topChar=st.peek();

                if((topChar=='('&&curr==')')||
                (topChar=='{'&&curr=='}')||
                (topChar=='['&&curr==']')){
                    st.pop();
                }else{
                    return false;
                }
            }
        }

        return st.isEmpty();
        
    }
}
// @lc code=end

